lib/engine/route.rb

```diff
@@ -59,56 +60,79 @@ def tail
       connection_data[-1]
     end
 
+    def chains
+      connection_data&.map { |c| c[:chain] }
+    end
+
     def connections
-      connection_data&.map { |c| c[:connection] }
+      chains
     end
 
-    def next_connection(node, connection, other)
-      connections = select(node, other, connection)
-      index = connections.find_index(connection) || connections.size
-      connections[index + 1]
+    def next_chain(node, chain, other)
+      chains = select(node, other, chain)
+      index = chains.find_index(chain) || chains.size
+      chains[index + 1]
     end
 
     def select(node, other, keep = nil)
       other_paths = @game.compute_other_paths(@routes, self)
 
       connection_data.each do |c|
-        next if c[:connection] == keep
+        next if c[:chain] == keep
 
-        other_paths.concat(c[:connection].paths)
+        other_paths.concat(c[:chain][:paths])
       end
 
-      nodes = [node, other]
+      get_node_chains(node, other).select { |c| (c[:paths] & other_paths).empty? }
+    end
 
-      node.hex.all_connections.select do |c|
-        (c.nodes & nodes).size == 2 && (c.paths & other_paths).empty?
-      end
+    # walk paths from start_node, and only keep connections that end at end_node
+    def get_node_chains(start_node, end_node)
+      @node_chains[[start_node, end_node]] ||=
```

Comment:
you never need to clear the cache?

Reply:
I don't think so, since it should never change based on user input.